File: java-algo/src/main/java/com/leetcode/Problem890.java
package com.leetcode;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 890. Find and Replace Pattern
 * https://leetcode.com/problems/find-and-replace-pattern/
 */
public class Problem890 {
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        List<String> ans = new ArrayList<>();
        for (String word : words) {
            if (check(word, pattern) && check(pattern, word)) {
                ans.add(word);
            }
        }
        return ans;
    }

    private boolean check(String word, String pattern) {
        if (word.length() != pattern.length()) {
            return false;
        }
        Map<Character, Character> map = new HashMap<>();
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i), p = pattern.charAt(i);
            if (map.containsKey(p)) {
                if (map.get(p) != c) {
                    return false;
                }
            } else {
                map.put(p, c);
            }
        }
        return true;
    }
}

File: python-algo/leetcode/problem_890.py
"""890.
"""
from typing import List


class Solution:
    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        def valid(w: str, p: str) -> bool:
            if len(w) != len(p):
                return False
            store1, store2 = dict(), dict()
            for i in range(len(w)):
                if p[i] in store1:
                    if store1[p[i]] != w[i]:
                        return False
                else:
                    store1[p[i]] = w[i]
                if w[i] in store2:
                    if store2[w[i]] != p[i]:
                        return False
                else:
                    store2[w[i]] = p[i]
            return True

        ans = []
        for word in words:
            if valid(word, pattern):
                ans.append(word)
        return ans
